crates/bounded_planet_core/src/networking/systems.rs

```diff
@@ -117,190 +123,186 @@ pub fn receive_net_events(
+            Ok(connection) => connection,
             Err(e) => {
-                event_sender.send(ReceiveEvent::NetworkError(
+                self.event_sender.send(ReceiveEvent::NetworkError(
                     NetworkError::ConnectionError(e)
                 )).expect("Failed to send network event");
-                break;
-            },
-            Ok(recv) => {
-                tokio::spawn(read_from_stream(id, recv, event_sender.clone()));
+                return;
             }
         };
-    }
 
-    // Send a final event indicating that this connection closed
-    event_sender
-        .send(ReceiveEvent::Disconnected(id))
-        .expect("Failed to send network event");
+        // Create a new MPSC which the ECS can use to send packets through this connection
+        let (send, recv) = unbounded_channel();
+
+        // Send an intial event indicating that this connection opened
+        self.event_sender
+            .send(ReceiveEvent::Connected(self.id, send))
+            .expect("Failed to send network event");
+
+        // Start running tasks to send/receive to this connection
+        tokio::spawn(Connected::new(
+            self.id,
+            connection,
+            uni_streams,
+            recv,
+            self.event_sender,
+        ).run());
+    }    
 }
 
-/// Handle all the work of a specific stream
-async fn read_from_stream(
-    connection_id: ConnectionId,
-    stream_recv: RecvStream<TlsSession>,
-    event_sender: UnboundedSender<ReceiveEvent>,
-) {
-    info!("Stream incoming: conn:{:?}", connection_id);
-
-    // Pull packets from this stream and publish them to the ECS through the event_sender
-    let mut recv = BoundedPlanetRecvStream::new(stream_recv);
-    loop {
-        let pkt = recv.recv_packet().await;
-        let se = match pkt {
-            Ok(pkt) => event_sender.send(ReceiveEvent::ReceivedPacket {
-                connection: connection_id,
-                data: Arc::new(pkt),
-            }),
-            Err(err) => {
-                event_sender.send(ReceiveEvent::NetworkError(
-                    NetworkError::ReceiveError {
-                        connection: connection_id,
-                        err,
-                    }
-                )).expect("Failed to send error event!");
+/// Represents an active quinn connection
+struct Connected {
+    id: ConnectionId,
+    connection: quinn::Connection,
+    uni_streams: IncomingUniStreams,
+    recv: UnboundedReceiver<SendEvent>,
+    send: UnboundedSender<ReceiveEvent>
+}
+
+impl Connected {
+    fn new(
+        id: ConnectionId,
+        connection: quinn::Connection,
+        uni_streams: IncomingUniStreams,
+        recv: UnboundedReceiver<SendEvent>,
+        send: UnboundedSender<ReceiveEvent>
+    ) -> Self {
+        Connected {
+            id,
+            connection,
+            uni_streams,
+            recv,
+            send
+        }
+    }
```

Comment:
This is actually kinda unidiomatic... You should just manually construct the struct where used, see other comment with suggestion for reference.